test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+            var t: Tensor(d.rank,eltType) = from;
+            return t;
+        }
+        
+        // ------------ Wasnt sure what to do with these   ----------------
+
+        /*operator =(ref lhs: Tensor(?rank,?eltType), in rhs: ?it) where (isRefIterType(it) || (isArray(rhs) && rhs.eltType == eltType)) && rhs.rank == rank {
+            lhs.reshapeDomain(rhs.domain);
+            lhs.data = rhs;
+        }
+        proc init=(in rhs: ?it) where isRefIterType(it) || isArray(rhs) {
+            this.init(rank,eltType);
+            this.reshapeDomain(rhs.domain);
+            this.data = rhs;
+        }
+        operator :(in from: ?it, type toType: Tensor(?rank,?eltType)) where (isRefIterType(it) || (isArray(from) && from.eltType == eltType)) && from.rank == rank {
+            // compilerError("Cannot convert from ",from.type:string," to ",toType:string);
+            var t: Tensor(rank,eltType) = from;
+            return t;
+        }*/
+
+        operator =(ref lhs: Tensor(?rank,?eltType), itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            lhs.reshapeDomain(itr.domain);
+            lhs.data = itr;
+        }
+
+        proc init=(itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            const A = itr;
+            this.init(A);
+        }
+
+        operator :(itr, type toType: Tensor(?rank,?eltType)) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            var t: Tensor(rank,eltType) = itr;
+            return t;
+        }
+
+
+        // Transposes a vector to a row matrix
+        proc transpose() where rank == 1 {
+            const (p,) = shape;
+            var t = new Tensor(eltType,1,p);
+            t.data[0,..] = this.data;
+            return t;
+        }
+
+        // Matrix transpose
+        proc transpose() where rank == 2 {
+            const (m,n) = this.shape;
+            var M = new Tensor(2,eltType);
+            M.reshapeDomain({0..#n,0..#m});
+            forall (i,j) in M.domain with (ref M, ref this) {
+                M.data[i,j] = this.data[j,i];
+            }
+            return M;
+        }
+
+        // Normalizes the tensor to have unit frobenius norm
+        proc normalize() {
+            const norm = sqrt(frobeniusNormPowTwo(this));
+            const data = this.data / norm;
+            return new Tensor(data);
+        }
+
+        // Retruns new tensor with provided domain
+        proc reshape(dom_) {
+            const dom = domainFromShape((...dom_.shape));
+            var t = new Tensor(dom.rank,eltType);
+            t.reshapeDomain(dom);
+            t.data = for (i,a) in zip(t.domain,this.data) do a;
+            return t;
+        }
+
+        // Returns new tensor with provided shape
+        proc reshape(shape: int ...?d) {
+            const dom = domainFromShape((...shape));
+            return this.reshape(dom);            
+        }
```

Comment:
I'm somewhat worried about the performance implication calling these `reshape` methods in a loop since they create a copy of the tensor that is typically only used temporarily. Would there be a way to refactor some of the code that uses this to avoid the copy?

For example, I'm thinking you could change the lines:
```chapel
const dL_dIn: Tensor(1) = dZ_dIn.transpose() * dL_dZ;
// ...
return dL_dIn.reshape(input.domain);
```

To call a new method that does all of the computation in a condensed fashion without creating temporary tensors:
```chapel
return dZ_dln.transposeMultiplyInto(dl_dZ, input.domain);
```